### PR TITLE
[WIP]xgboost4j-spark

### PR DESCRIPTION
**Pull request by CodingCat**
it is a WIP PR for xgboost4j-spark

### CODE REVIEW

**Comment by CodingCat** [comment]
@tqchen , one thing we need to confirm as how the format of DMatrix binary file looks like, 

In Spark/Hadoop, the output looks like 

```
dmatrix1-partition0

dmatrix1-partition1

dmatrix1-partition2

```

Can we assume each partition forms a complete DMatrix?

**Comment by tqchen** [comment]
@CodingCat 
- I think we do not need to save DMatrix binary file, instead always load data from spark's native data structure
- Please remove tracker.py from repo as it is created by copy from dmlc-core/tracker/dmlc_tracker. It would be great if we have a pom that does this copy in build time.
- Since we are already under package(namespace spark). I think we are safe to name the class XGBoost as opposed to XGBoostSpark

**Comment by tqchen** [comment]
in another words, in the spark and other dataflow interface, the input data is RDD, DataSet and DMatrix is like developer API that is invisible in the higher level

**Comment by tqchen** [comment]
please do a squash of commits, so the file creation and deletion won't get into history

File: jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    val nFold = config.getInt("nfold")
+    // TODO: build configuration map from config
+    val xgBoostConfigMap = new HashMap[String, AnyRef]()
+    sc.binaryFiles(filePath, numWorkers).mapPartitions{
+      partitionIterator =>
+        val results = new ListBuffer[String]
+        for ((_, fileInStream) <- partitionIterator) {
+          val dMatrix = DMatrixBuilder.buildDMatrixfromBinaryData(fileInStream.toArray())
+          results.insertAll(results.size,
+            XGBoost.crossValidation(xgBoostConfigMap, dMatrix, round, nFold, null, obj, eval))
```
Comment (tqchen) [inline jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: let us not yet do crossValidation, cv is not yet supported in current API, only do distributed training

**Comment by tqchen** [comment]
- The first step goal is to make a training API that load data distributedly like DistTrain to validate the concept. 
- Then we want to take RDD[LabeledPoint] in sparkMLLib as input, and convert that into DMatrix

**Comment by CodingCat** [comment]
hmmm....

Let me describe with an example 

case 1: 

we have a training dataset which is a larger version of `demo/data/agaricus.txt.train`, we save it in HDFS and load as RDD[String]; for each partition, it is a Iterator[String].


case 2: dense matrix, 

the RDD loaded from HDFS is essentailly RDD[Float], for each partition, it is Iterator[Float], 

case 3: sparse matrix (CSR)

In this case, we just need to ensure that we partition the data in a perfect way, so that each partition of RDD is a `parse-able` CSR format, the partition of the RDD loaded from HDFS is Iterator[CSRPartition]


In the above case, we need to ensure that when we save file, each partition of the file is parse-able, especially for the third case, right?

File: jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/Boosters.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import org.apache.spark.rdd.RDD
+
+class Boosters(boosters: RDD[Booster]) {
+
+  def save(path: String): Unit = {
+
+  }
+
+  def chooseBestBooster(boosters: RDD[Booster]): Booster = {
+    // TODO:
```
Comment (tqchen) [inline jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/Boosters.scala]: I think we only need Booster here

Comment (tqchen) [inline jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/Boosters.scala]: or GradientBooster

File: jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  private def buildSparkContext(config: Config): SparkContext = {
+    if (_sc.isEmpty) {
+      // TODO:build SparkContext with the user configuration (cores per task, and cores per executor
+      // (or total cores)
+      // NOTE: currently Spark has limited support of configuration of core number in executors
+    }
+    _sc.get
+  }
+
+  def train(config: Config, obj: ObjectiveTrait = null, eval: EvalTrait = null): RDD[Booster] = {
```
Comment (tqchen) [inline jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: User should not see RDD, the return value is a Booster, that contains  result from RDD

File: jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/DMatrixTest.java [hunk 4]
```diff
@@ -29,41 +29,6 @@
 public class DMatrixTest {
 
   @Test
-  public void testCreateFromDataIterator() throws XGBoostError {
```
Comment (tqchen) [inline jvm-packages/xgboost4j/src/test/java/ml/dmlc/xgboost4j/DMatrixTest.java]: This testcase is removed in your PR

File: jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import ml.dmlc.xgboost4j.{DMatrix => JDMatrix}
+import ml.dmlc.xgboost4j.scala.{DMatrix, Booster}
+import org.apache.spark.mllib.regression.LabeledPoint
+import org.apache.spark.rdd.RDD
+
+class XGBoostModel(booster: Booster) extends Serializable {
+
+  def predict(testSet: RDD[LabeledPoint]): RDD[Array[Array[Float]]] = {
+    val broadcastBooster = testSet.sparkContext.broadcast(booster)
```
Comment (tqchen) [inline jvm-packages/xgboost4jspark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: check what does MLLib returns, probably a matrix

**Pull request merged**